onnxruntime/core/providers/nnapi/nnapi_builtin/nnapi_execution_provider.cc

```diff
@@ -0,0 +1,356 @@
+                                                   [](int) { return onnxruntime::make_unique<CPUAllocator>(onnxruntime::make_unique<OrtMemoryInfo>(NNAPI, OrtAllocatorType::OrtDeviceAllocator, OrtDevice(), 0, OrtMemTypeCPUOutput)); },
+                                                   std::numeric_limits<size_t>::max()});
+
+  InsertAllocator(CreateAllocator(cpu_memory_info));
+}
+
+NnapiExecutionProvider::~NnapiExecutionProvider() {}
+
+std::vector<std::unique_ptr<ComputeCapability>>
+NnapiExecutionProvider::GetCapability(const onnxruntime::GraphViewer& graph,
+                                      const std::vector<const KernelRegistry*>& /*kernel_registries*/) const {
+  std::vector<std::unique_ptr<ComputeCapability>> result;
+  // Handle If and Loop operators
+  if (graph.IsSubgraph()) {
+    return result;
+  }
+
+  // Need access to model_path_
+  for (const auto& tensor : graph.GetAllInitializedTensors()) {
+    if (tensor.second->has_data_location() &&
+        tensor.second->data_location() == ONNX_NAMESPACE::TensorProto_DataLocation_EXTERNAL) {
+      LOGS_DEFAULT(WARNING) << "NNAPI: Initializers with external data"
+                               " location are not currently supported";
+      return result;
+    }
+  }
+
+  // TODO, switch to use graph instead of model
+  // This method is based on that of TRT EP
+  // Construct modelproto from graph
+  onnxruntime::Model model(graph.Name(), true, ModelMetaData(),
+                           PathString(),
+                           IOnnxRuntimeOpSchemaRegistryList(),
+                           graph.DomainToVersionMap(),
+                           std::vector<ONNX_NAMESPACE::FunctionProto>(),
+                           *GetLogger());
+  std::unordered_set<std::string> all_node_inputs;
+  onnxruntime::Graph& graph_build = model.MainGraph();
+  for (const auto& node : graph.Nodes()) {
+    std::vector<onnxruntime::NodeArg*> inputs, outputs;
+    for (auto input : node.InputDefs()) {
+      auto& n_input = graph_build.GetOrCreateNodeArg(input->Name(), input->TypeAsProto());
+      inputs.push_back(&n_input);
+      all_node_inputs.insert(input->Name());
+    }
+    for (auto output : node.OutputDefs()) {
+      auto& n_output = graph_build.GetOrCreateNodeArg(output->Name(), output->TypeAsProto());
+      outputs.push_back(&n_output);
+    }
+    graph_build.AddNode(node.Name(), node.OpType(), node.Description(), inputs, outputs, &node.GetAttributes(), node.Domain());
+  }
+  //Add initializer to graph
+  const auto& init_tensors = graph.GetAllInitializedTensors();
+  for (const auto& tensor : init_tensors) {
+    graph_build.AddInitializedTensor(*(tensor.second));
+  }
+
+  ORT_ENFORCE(graph_build.Resolve().IsOK());
+  ONNX_NAMESPACE::ModelProto model_proto = model.ToProto();
+  model_proto.set_ir_version(ONNX_NAMESPACE::Version::IR_VERSION);
+
+  nnapi::ModelBuilder builder(model_proto);
+  const auto supported_nodes_vector = builder.GetSupportedNodes();
```

Comment:
Still not quite getting the need for this. You create a new Model and build a new Graph within that in order to get a ModelProto to pass to ModelBuilder. But ModelBuilder only seems to use the GraphProto from ModelProto.graph(). 

Can ModelBuilder take a GraphProto as input instead of ModelProto? 

If so you can just use the original GraphProto via graph.GetGraph().ToGraphProto() (would be less confusing if the GraphViewer arg is renamed to graph_viewer though).

Reply:
This is a TODO to switch entirely to use the Graph instead of ModelProto, the reason I did not use the  graph.GetGraph().ToGraphProto() and keep the existing impl is that the  graph.GetGraph().ToGraphProto()  does not contain the value_info, while this can be resolved when I switch to using Graph